lib/assert.js

```diff
@@ -22,18 +22,32 @@ class AssertionError extends Error {
 		// Reserved for power-assert statements
 		this.statements = [];
 
+		// To keep compatibility with old API, don't know if needed
 		if (opts.stack) {
 			this.stack = opts.stack;
+		} else if (opts.callSite) {
+			this.stack = opts.callSite.getStack(this.toString());
 		}
 	}
 }
 exports.AssertionError = AssertionError;
 
-function getStack() {
-	const obj = {};
-	Error.captureStackTrace(obj, getStack);
-	return obj.stack;
+class CallSite {
+	constructor(caller) {
+		// This is not lazy in node. However, using pure `Error` instances might not be
+		// backwards compatible (nor will it allow us to remove unwanted frames easily).
+		Error.captureStackTrace(this, caller || CallSite);
+	}
+
+	toString() {
+		return '[[CallSite]]';
+	}
+
+	getStack(target) {
+		return this.stack.replace('[[CallSite]]', target || '');
```

Comment:
You're correct. When you call `Error.captureStackTrace(obj, fn)`, it will produce `obj.toString()\n<stack trace>`. I simply inserted a token that should not happen during any stacktrace so that it could be replaced at a later time.